File: rosWorkspace/PathTask/src/PathTask.cpp
#include "PathTask.hpp"
#include "Robosub/HighLevelControl.h"
#include <stdio.h>
#include <math.h>
#include <float.h>

/**
 * @brief The constructor. Subscribes to all needed topics and sets up callbacks
 */
PathTask::PathTask()
 : m_nodeHandle(),
   m_pathSubscriber(),
   m_taskStateSubscriber(),
   m_pathDirSubscriber(),
   m_highLevelMotorPublisher(),
   m_taskCompletePublisher(),
   m_isEnabled(false),
   m_frames(0),
   m_direction("left"),
   m_msgList(),
   m_lastId(-1)
{
   printf("setup\n");
   fflush(NULL);
   m_pathSubscriber = m_nodeHandle.subscribe("img_rec/paths", 10, &PathTask::pathCallback, this);
   m_taskStateSubscriber = m_nodeHandle.subscribe("Module_Enable", 10, &PathTask::moduleEnableCallback, this);
   m_pathDirSubscriber = m_nodeHandle.subscribe("Path_Direction", 10, &PathTask::pathDirectionCallback, this);
   m_highLevelMotorPublisher = m_nodeHandle.advertise<Robosub::HighLevelControl>("High_Level_Motion", 10);
   m_taskCompletePublisher = m_nodeHandle.advertise<std_msgs::String>("Task_Completion", 10);
}

/**
 * @brief The destructor
 */
PathTask::~PathTask()
{

}

/**
 * @brief The module enable message callback
 *
 * @param msg The module enabled message
 */
void PathTask::moduleEnableCallback(const Robosub::ModuleEnableMsg& msg)
{
  if (msg.Module == "PathTask")
  {
    if (msg.State == true)
    {
      printf("PathTask: enabled\n");
      m_isEnabled = true;
      m_frames = 0;
      m_msgList.clear();
    }
    else
    {
      printf("PathTask: Disabled\n");
      m_isEnabled = false;
    }
  }
}

void PathTask::pathDirectionCallback(const std_msgs::String& msg)
{
  m_direction = msg.data;
}

/**
 * @brief The path callback
 *
 * @param msg The image recognition object message
 */
void PathTask::pathCallback(const SubImageRecognition::ImgRecObject& msg)
{
	if (m_isEnabled)
	{
	  if (msg.id == 0 && m_lastId != -1)
	  {
	    //find the correct target
	    SubImageRecognition::ImgRecObject tmp = getMessage();

	    manualMode(msg);
	    //offsetMode(msg);
	    m_msgList.clear();
	    m_msgList.push_back(msg);
	  }
	  else
	  {
	    m_msgList.push_back(msg);
	  }

    m_lastId = msg.id;
	}
}

void PathTask::manualMode(const SubImageRecognition::ImgRecObject& msg)
{
  int hit = 0;
  float moveX = msg.center_x;
  float moveY = msg.center_y;
  float turn = msg.rotation;

  if (moveX > 10 || moveX < -10)
  {
    printf("Correcting straf by: %f\n", moveX);
    publishMotor("Straf", "Manual", moveX);
    hit++;
  }
  else
  {
    publishMotor("Straf", "Manual", 0.0);
  }

  if (moveY > 10 || moveY < -10)
  {
    printf("Correcting forward by: %f\n", moveY);
    publishMotor("Forward", "Manual", moveY*2.0);
    hit++;
  }
  else
  {
    publishMotor("Forward", "Manual", 0.0);
  }

  if (turn > 1 || turn < -1)
  {
    printf("Correcting turn by: %f\n", turn);
    publishMotor("Turn", "Manual", turn);
    hit++;
  }
  else
  {
    publishMotor("Turn", "Manual", 0.0);
  }

  if (hit == 0)
  {
    m_frames++;
  }
  else
  {
    m_frames = 0;
  }

  if (m_frames >= 10)
  {
    reportSuccess(true);
    m_frames = 0;
    m_isEnabled = false;
  }
}

void PathTask::offsetMode(const SubImageRecognition::ImgRecObject& msg)
{
  int hit = 0;
  float moveX = calculateDistanceFromCenter(msg.center_x, msg.width);
  float moveY = calculateDistanceFromCenter(msg.center_y, msg.width);
  float turn = msg.rotation;

  if (moveX > 10 || moveX < -10)
  {
    printf("Correcting straf by: %f\n", moveX);
    publishMotor("Straf", "Offset", moveX > 0 ? 1 : -1); //inches to feet
    hit++;
  }
  else
  {
    publishMotor("Straf", "Offset", 0.0);
  }

  if (moveY > 10 || moveY < -10)
  {
    printf("Correcting forward by: %f\n", moveY);
    publishMotor("Forward", "Offset", moveY > 0 ? 1 : -1);
    hit++;
  }
  else
  {
    publishMotor("Forward", "Offset", 0.0);
  }

  if (turn > 1 || turn < -1)
  {
    printf("Correcting turn by: %f\n", turn);
    publishMotor("Turn", "Offset", turn);
    hit++;
  }
  else
  {
    publishMotor("Turn", "Offset", 0.0);
  }

  if (hit == 0)
  {
    m_frames++;
  }
  else
  {
    m_frames = 0;
  }

  if (m_frames >= 10)
  {
    reportSuccess(true);
    m_frames = 0;
    m_isEnabled = false;
  }
}

SubImageRecognition::ImgRecObject PathTask::getMessage()
{
  SubImageRecognition::ImgRecObject ret;

  if (m_direction == "right")
  {
    int greatest = SHRT_MIN;
    for (std::list<SubImageRecognition::ImgRecObject>::iterator it = m_msgList.begin(); it != m_msgList.end(); it++)
    {
      if (it->center_x > greatest)
      {
        greatest = it->center_x;
        ret = *it;
      }
    }
  }
  else if (m_direction == "left")
  {
    int greatest = SHRT_MAX;
    for (std::list<SubImageRecognition::ImgRecObject>::iterator it = m_msgList.begin(); it != m_msgList.end(); it++)
    {
      if (it->center_x < greatest)
      {
        greatest = it->center_x;
        ret = *it;
      }
    }
  }
  else
  {
    if (m_msgList.size() > 0)
    {
      ret = m_msgList.front();
    }
  }

  return ret;
}

/**
 * @brief report the success or failure of the module
 *
 * @param success The success or failure
 */
void PathTask::reportSuccess(bool success)
{
  std_msgs::String msg;
  msg.data = "PathTask";
  if (success)
  {
    msg.data += " Success";
  }
  else
  {
    msg.data = " Failure";
  }
  m_taskCompletePublisher.publish(msg);
}

/**
 * @brief plubish to the high level motor controller topic
 *
 * @param direction The direction: Forward, Straf, Turn
 * @param motion The motion type: Manual, Offset
 * @param value The value
 */
void PathTask::publishMotor(std::string direction, std::string motion, float value)
{
  Robosub::HighLevelControl msg;
  msg.Direction = direction;
  msg.MotionType = motion;
  msg.Value = value;

  m_highLevelMotorPublisher.publish(msg);
}

/**
 * @brief Calcualte the distance the center of the object is off from the center of the camera in inches
 *
 * @param centerDir The current center direction
 * @param width The width of the object
 *
 * @return The distance in inches
 */
float PathTask::calculateDistanceFromCenter(float centerDir, float width)
{
  float pxlPerInch = getPixelsPerInch(width, 6.0f);
  return centerDir * pxlPerInch;
}

/**
 * @brief Get the number of pixels per inch
 *
 * @param curWidthPixels The current width of a known object in pixels
 * @param expectedWidthInches The expected with of an object in inches
 *
 * @return The number of pixels per inch if everything goes well, or float_max if inputs were unacceptable
 */
float PathTask::getPixelsPerInch(float curWidthPixels, float expectedWidthInches)
{
  float dist = getDistance(curWidthPixels, expectedWidthInches);
  return curWidthPixels/dist;
}

/**
 * @brief Calculate the distance to the object
 *
 * @param curObjSize The objects current size
 * @param actualObjSize The objects expected size
 *
 * @return The distance
 */
float PathTask::getDistance(float curObjSize, float actualObjSize)
{
  return (actualObjSize/(2.0f*tan(curObjSize * (M_PI/4.0f)/960.0f))); //45 degrees
}

File: rosWorkspace/PathTask/src/PathTask.py
#!/usr/bin/python

import roslib
roslib.load_manifest('PathTask')
import rospy
from std_msgs.msg import String
import time
from threading import Thread


from Robosub.msg import HighLevelControl, ModuleEnableMsg
from SubImageRecognition.msg import ImgRecObject


class PathTask:
    
    MOTOR_COMMAND = 'Command'
    MOTOR_FORWARD = 'Forward'
    MOTOR_MANUAL = 'Manual'
    MOTOR_OFFSET = 'Offset'
    MOTOR_STRAFE = 'Straf'
    MOTOR_TURN = 'Turn'
    
    SCALE_FORWARD = 0.0017
    SCALE_STRAFE = 0.01
    SCALE_TURN = 1 / 180.0
    
    SUCCESS_GOAL = 10
    
    TH_ROT = 1
    TH_X = 50
    TH_Y = 50
    
    def __init__(self):
        self.can_turn = True
        self.direction = 'right'
        self.enabled = False
        self.last_motor_change = 0
        self.paths = []
        self.pub_high_level_motor_controller = rospy.Publisher(
                'High_Level_Motion', HighLevelControl)
        self.pub_task_complete = rospy.Publisher(
                'Task_Completion', String)
        self.sub_image_recognition = rospy.Subscriber(
                'img_rec/paths', ImgRecObject, self.image_recognition_cb)
        self.sub_module_enable = rospy.Subscriber(
                'Module_Enable', ModuleEnableMsg, self.module_enable_cb)
        self.sub_path_direction = rospy.Subscriber(
                'Path_Direction', String, self.path_direction_cb)
        self.success_counter = 0
        self.thread = Thread(target=self.motor_watcher)
        self.thread.daemon = True
        self.thread.start()
    
    def align_to_path(self, path):
        did_something = False
        if path.center_x > self.TH_X or path.center_x < -self.TH_X:
            self.publish_motor(self.MOTOR_STRAFE, path.center_x * self.SCALE_STRAFE)
            did_something |= True
        else:
            self.publish_motor(self.MOTOR_STRAFE, 0)
        if path.center_y > self.TH_Y or path.center_y < -self.TH_Y:
            self.publish_motor(self.MOTOR_FORWARD, path.center_y * self.SCALE_FORWARD)
            did_something |= True
        else:
            self.publish_motor(self.MOTOR_FORWARD, 0)
        if self.can_turn and (path.rotation > self.TH_ROT or path.rotation < -self.TH_ROT):
            self.publish_motor(self.MOTOR_TURN, path.rotation * self.SCALE_TURN)
            did_something |= True
        else:
            self.publish_motor(self.MOTOR_TURN, 0)
        if did_something:
            self.success_counter = 0
            self.last_motor_change = time.time()
        else:
            self.success_counter += 1
            if self.success_counter >= self.SUCCESS_GOAL:
                if self.can_turn:
                    self.task_complete(True)
                else:
                    self.can_turn = True
                    self.success_counter = 0
    
    def image_recognition_cb(self, path):
        if not self.enabled:
            return
        if len(self.paths) and path.id == 0:
            self.align_to_path(self.select_correct_path())
            self.paths = []
        self.paths.append(path)
    
    def module_enable_cb(self, msg):
        if msg.Module == 'PathTask':
            self.can_turn = False
            self.enabled = msg.State
            self.paths = []
            self.success_counter = 0
            self.last_motor_change = 0
            if not self.enabled:
                self.stop_motors()
                rospy.loginfo("PathTask Disabled")
            else:
                rospy.loginfo("PathTask Enabled")
    
    def motor_watcher(self):
        while True:
            if self.enabled and self.last_motor_change:
                if time.time() - self.last_motor_change > 3:
                    self.task_complete(False)
            time.sleep(1)
    
    def path_direction_cb(self, msg):
        if msg.data in ('left', 'right'):
            self.direction = msg.data
        else:
            print('[PathTask] Invalid path direction received: ' + msg.data)
    
    def publish_motor(self, direction, value, motion_type=None):
        if not motion_type:
            if direction == self.MOTOR_TURN:
                motion_type = self.MOTOR_MANUAL
            else:
                motion_type = self.MOTOR_COMMAND
        msg = HighLevelControl()
        msg.Direction = direction
        msg.Value = value
        msg.MotionType = motion_type
        self.pub_high_level_motor_controller.publish(msg)
    
    def select_correct_path(self):
        if len(self.paths) == 1:
            return self.paths[0]
        elif self.direction == 'left':
            best = (9999, None)
            for path in self.paths:
                if path.center_x < best[0]:
                    best = (path.center_x, path)
        else:
            best = (-9999, None)
            for path in self.paths:
                if path.center_x > best[0]:
                    best = (path.center_x, path)
        return best
    
    def task_complete(self, result):
        self.enabled = False
        result = 'PathTask ' + 'Success' if result else 'Failure'
        self.pub_task_complete.publish(String(result))
        self.stop_motors()
    
    def stop_motors(self):
        self.publish_motor(self.MOTOR_FORWARD, 0)
        self.publish_motor(self.MOTOR_STRAFE, 0)
        self.publish_motor(self.MOTOR_TURN,  0)
        self.last_motor_change = 0


if __name__ == '__main__':
    rospy.init_node('PathTask')
    path_task = PathTask()
    rospy.spin()

